### DISCUSSION TITLE
(CSS, React, HeadlessUI) Show overflow of a deeply nested tooltip when grand-grandparent has `overflow-hidden`

### DISCUSSION THREAD

**Discussion by bennettdams** [tailwindlabs/tailwindcss#10363] [Help]
Hey!
My question is more related to CSS, but also applies to HeadlessUI's `Popover` component.

I have a reusable `Card` component all over the place that just brings some styling and render `children`. This component has `overflow-hidden`. 

Now I want to show a tooltip at any nested level of the children - imagine the `Card` renders a multiple `div`s until finally the tooltip is rendered. The tooltip can be bigger than the `Card` it is rendered in.

In this scenario, what Is the correct way for the tooltip to break out of the grand-grand-grandparent's `overflow-hidden`? There has to be a way with CSS/React portals, I just don't know it :sweat_smile:

Please remember that I don't control the `Card`'s overflow because me tooltip is rendered somewhere deeply nested inside. I also don't control the `Card`'s parent to [apply the `position: relative` trick](https://css-tricks.com/popping-hidden-overflow/), because we're deeply nested in the component structure.

- Is there a way to do this with plain Tailwind CSS utilities?
- Is there a way to do this with HeadlessUI's `Popover`?
- Do you have to use a React portal maybe?

Here is a Playground example:

https://play.tailwindcss.com/fWMTPVXbKA

This is what I want to achieve with the tooltip, it (green box) overflows the card (blue box). Here, the card has no `overflow-hidden` for demonstration purposes:

![image](https://user-images.githubusercontent.com/29319414/213454921-d2a3a8cd-fcd7-4874-bfeb-bc28df73529d.png)

In reality, my card does have `overflow-hidden`. The tooltip (green box) is cut off at the card boundary (blue box):

![image](https://user-images.githubusercontent.com/29319414/213454990-74df5aaf-3aa9-4908-b9ec-87408754823c.png)

**Comment by wongjn** [comment]
> In this scenario, what Is the correct way for the tooltip to break out of the grand-grand-grandparent's `overflow-hidden`?

Use a portal for the tooltip element.

> Is there a way to do this with plain Tailwind CSS utilities?

Nope!

> Is there a way to do this with HeadlessUI's `Popover`?

Not out-of-the-box.

> Do you have to use a React portal maybe?

Yep!

**Reply by bennettdams** [reply]
Thanks! 

> Use a portal for the tooltip element.

Do you have an example for that?

When using a portal for this scenario, the portal root needs to sit outside of my `Card` component. As described above, I can be multiple nesting levels inside a `Card`, so the only way I can think of is a portal root at the root of the app itself. Somewhere next to `<body />` or something. This means that I need...
- to calculate the tooltip position somehow programmatically based on the next sibling or parent wherever the tooltip hover element is living
- to have multiple portal roots on the body with different IDs, as I can have an arbitrary amount of tooltips in my app, because every tooltip needs its own portal root, right?

**Reply by wongjn** [reply]
>> Use a portal for the tooltip element.
>
> Do you have an example for that?

The [React documentation on Portals](https://reactjs.org/docs/portals.html) has several examples.
